### ISSUE TITLE
How to scale

### ISSUE THREAD

**Issue by Bongsakorn** [pinterest/secor#182]
One great features of Secor is **horizontal scalability**. How the secor can scale?

When I compiled the secor-0.7.tar.gz file. I unpack the archive and run the service. So If I'd like to scale, I just copy the secor-0.7.tar.gz file to another machine and run service. right?

**Comment by HenryCaiHaiying** [comment]
You can copy that jar to as many machines as you want and start the secor
java process on those machines.  These secor processes on different
machines will join the same kafka high level consumer group.  The high
level consumer code would distribute the kafka topic/partitions evenly
between these java processes, it also takes care of the situation when some
processes died during the middle or new process joined.

**Comment by abstract-karshit** [comment]
@HenryCaiHaiying Doesn't it depend on the number of partitions a particular topic has in Kafka?

**Comment by HenryCaiHaiying** [comment]
Yes, you would need to have multiple partitions on kafka broker first.
Secor is the consumer for those topics, you can have one secor node to
process multiple kafka partitions.  But if you realize that one secor node
is not able to catch up with the kafka data volume in that topic, you would
need to create multiple secor nodes/processes to distribute the work load.

**Comment by abstract-karshit** [comment]
Thanks @HenryCaiHaiying . So if my Topic has 15 Kafka partitions and one secor process is running 5 threads, in an ideal case I should run 3 processes (5 threads each) on different nodes to consume data from all 15 partitions via zookeeper?
Or, does spawning the process to more machines i.e. running more threads in total than the number of topic partitions help in any scenario?

**Comment by HenryCaiHaiying** [comment]
One thread can pull multiple kafka partitions, really depends on the volume
in your partition and the node's CPU/memory/network bandwidth.  You can try
one node first (with 5 threads) to see if you hit network/CPU upper bound.

Running more threads than the number of partitions wouldn't help. some
threads is just going to do nothing.

**Comment by RainbowTraveller** [comment]
As a sub or supplementary question for this. I was able to achieve the horizontal scaling as per the above solution. I have 6 processes running each of which has 10 threads thus taking care of 60 partitions for a particular topic.

The 6 processes are divided as 2 + 4 on 2 different hosts. But I was unable to achieve log separation for each process. I want to configure logs for each process separately. Each of the process should log into separate file. Right now somehow all the processes are following some default config and ignoring the log4j config files.

**Comment by HenryCaiHaiying** [comment]
Is the problem on log4j config or secor config?

Secor config is specified through -Dconfig=<your-config-file>, did you
specify the same file for all processes?

**Comment by RainbowTraveller** [comment]
I have separate directories containing individual property files as well as log4j config file along with  secor jar.

**Comment by abstract-karshit** [comment]
@HenryCaiHaiying Thanks for the advice. I ran secor for same consumer group ID in three m4.xlarge machines (4 cores, 16 gb) running 5 threads each (so total 15 threads to read from total 15 partitions for my topic). In exact three hours I was able to upload ~86 million events to S3 (Json events in sequence file format using bzip2 compresion). So, current rate for my setup is : ~28 million events per hour.
Is there a way to increase it to 100 million events per hour. Primarily I need help in what properties to change in config file (currently set to defaults) to increase ingestion rate from Kafka and upload to S3 with sequence file and bzip2 compression.
Thanks
^^ @RainbowTraveller @pgarbacki @spektom Any help?

**Comment by spektom** [comment]
@abstract-karshit 

You should first understand where your bottleneck is. There are several options:

**Kafka**

In this case increasing number of partitions might help. We have topics with 120 partitions (these topics transfer more than ~1B of events, which is ~500GB of compressed JSON data daily).

**Network**

If you're running on Amazon, pay an attention that "smaller" instances have worse network throughput. Sometimes, upgrading to a bigger instance type improves things dramatically.

**CPU**

Secor processes can be CPU intensive, thus it maybe worth trying to run them on machines with greater number of virtual cores.

How to identify your bottleneck?

1. Use monitoring tools: top (CPU), iftop (Network).
2. Try scaling out one of the layers: Kafka or Secor cluster, and see how things change.

Here's our configuration, just for reference:

1. About ~10TB of events uploaded by Secor every day. 
2. 18 EC2 instances ranging from c3.xlarge to c4.2xlarge and m3.2xlarge.
3. We use Kafka lag as a metric, and in case it grows we scale relevant Secor cluster by adding more machines to it (one cluster processes several Kafka topics).
4. We monitor Secor machines for CPU usage, and in case it goes down for some reason we restart Secor service on an affected machine. Sometimes this happens because of some Kafka re-balancing issues (we use Kafka 0.8.x)

Hope this helps.

**Comment by abstract-karshit** [comment]
@spektom Thanks for your suggestion, will monitor all metrics and scale layers accordingly.

**Comment by HenryCaiHaiying** [comment]
Thanks @spektom, these are very detailed information.

In our experience, the bottleneck usually come from network bandwidth
constraint.  secor's cpu usage is mostly coming from deserializing json
message (to find the timestamp content) and do compression (if you use
bzip2 or gzip compression), 'top' command or other cpu monitoring tool can
roughly tells you how busy the machine is, try to make cpu usage under
60%.  Have each secor thread take one or two kafka partitions will help,
partition kafka topics into more smaller partitions does help too.

**Comment by abstract-karshit** [comment]
Thank you guys. @HenryCaiHaiying @spektom I am able to double my current ingestion rate and will try more variations tomorrow to scale further.

**Comment by abstract-karshit** [comment]
Hi @HenryCaiHaiying @spektom 
So, I am able to achieve my desired scale, thank to you guys. What I have seen is that after 2 hours of continuous running my application performance of Secor degrades drastically (in terms of CPU utilisation and Network in). When I see log, is is happening because of I see very frequent: 
**(kafka.utils.Logging$class:103) ERROR error during syncedRebalance kafka.common.ConsumerRebalanceFailedException: can't rebalance after 4 retries**
I am on Kafka 0.8.1.1 and my topic has 15 partitions. Currently running 5 m4.2xlarge machines with three threads each.

**Comment by spektom** [comment]
Hello @abstract-karshit,

You can try increasing these Kafka parameters: **rebalance.backoff.ms** and **zookeeper.session.timeout.ms**, and see if it helps. In any case, check that your Zookeeper cluster is not overloaded as it might cause these kind of issues.

**Comment by RainbowTraveller** [comment]
I have posted quite a few issues on this forum. What I was trying to do was scale up secor to suit the production needs. I was riddled with problems like checksum exception (HadoopS3UploadManager) and FileNotFoundException to compare/compute MD5 checksum or Data Stream ended unexpectedly(S3UploadManager). In this scenario I was running 5 processes on 2 machines as 2(m3.xlarge) + 3(c3.2xlarge). Each of which was having 10 threads. But I was unable to run this setup continuously and processes invariably died due to one of the above mentioned exceptions.

Then I moved all the 5 processes to one machine(c3.2xlarge), which at least gives me continuous running of processes unlike previous experience. But with usage of S3UploadManager (which may be the culprit, I am not sure), the resident memory per processes grows from 1.5G to 3.2G and eventually one of the processes if killed. I am now trying HadoopS3UploadManager and keeping fingers crossed. 

@spektom  : Are there any more settings needed to distribute the secor processes across multiple machines(keeping kafka consumer group same)? I am using AWS instances as mentioned above.
